L1Trigger/Phase2L1GT/plugins/L1GTProducer.cc

```diff
@@ -0,0 +1,452 @@
+    produces<P2GTCandidateCollection>("CL2Photons");
+    produces<P2GTCandidateCollection>("CL2Electrons");
+    produces<P2GTCandidateCollection>("CL2Taus");
+    produces<P2GTCandidateCollection>("CL2EtSum");
+    produces<P2GTCandidateCollection>("CL2HtSum");
+  }
+
+  void L1GTProducer::fillDescriptions(edm::ConfigurationDescriptions &description) {
+    edm::ParameterSetDescription desc;
+
+    edm::ParameterSetDescription scalesDesc;
+    L1GTScales::fillDescriptions(scalesDesc);
+    desc.add<edm::ParameterSetDescription>("scales", scalesDesc);
+
+    desc.addOptional<edm::InputTag>("GTTPromptJets");
+    desc.addOptional<edm::InputTag>("GTTDisplacedJets");
+    desc.addOptional<edm::InputTag>("GTTPrimaryVert");
+
+    desc.addOptional<edm::InputTag>("GMTSaPromptMuons");
+    desc.addOptional<edm::InputTag>("GMTSaDisplacedMuons");
+    desc.addOptional<edm::InputTag>("GMTTkMuons");
+
+    desc.addOptional<edm::InputTag>("CL2Jets");
+    desc.addOptional<edm::InputTag>("CL2Photons");
+    desc.addOptional<edm::InputTag>("CL2Electrons");
+    desc.addOptional<edm::InputTag>("CL2Taus");
+    desc.addOptional<edm::InputTag>("CL2EtSum");
+    desc.addOptional<edm::InputTag>("CL2HtSum");
+
+    description.addWithDefaultLabel(desc);
+  }
+
+  void L1GTProducer::produceGTTPrimaryVert(edm::Event &event) const {
+    std::unique_ptr<P2GTCandidateCollection> outputCollection = std::make_unique<P2GTCandidateCollection>();
+    const VertexWordCollection &collection = event.get(gttPrimaryVertexToken_);
+    for (std::size_t i = 0; i < collection.size() && i < 10; i++) {
+      const VertexWord &obj = collection[i];
+      P2GTCandidate gtObj;
+      gtObj.hwZ0_ = obj.z0Word().V.to_int() * 5;
+      gtObj.hwQual_ = obj.qualityWord().V.to_int();
+      gtObj.hwSum_pT_pv_ = obj.multiplicityWord().V.to_int();
+      gtObj.hwNumber_of_tracks_in_pv_ = obj.multiplicityWord().V.to_int();
+      gtObj.hwNumber_of_tracks_not_in_pv_ = obj.inverseMultiplicityWord().V.to_int();
+      gtObj.objectType_ = P2GTCandidate::GTTPrimaryVert;
+
+      gtObj.z0_ = scales_.to_z0(gtObj.hwZ0());
+
+      outputCollection->push_back(gtObj);
+    }
+    event.put(std::move(outputCollection), "GTTPrimaryVert");
+  }
+
+  void L1GTProducer::produceGTTPromptJets(edm::Event &event) const {
+    // static_assert(std::is_same<std::result_of<decltype (&TkJetWord::ptWord)(TkJetWord *)>::type,
+    //                            ap_ufixed<16, 11, AP_TRN, AP_SAT>>::value);
+    // static_assert(std::result_of<decltype (&TkJetWord::ptWord)(TkJetWord *)>::type::width == 16);
+    // static_assert(std::result_of<decltype (&TkJetWord::glbPhiWord)(TkJetWord *)>::type::width == 13);
+    // static_assert(std::result_of<decltype (&TkJetWord::glbEtaWord)(TkJetWord *)>::type::width == 14);
+    // static_assert(std::result_of<decltype (&TkJetWord::z0Word)(TkJetWord *)>::type::width == 10);
+    // static_assert(std::result_of<decltype (&TkJetWord::ntWord)(TkJetWord *)>::type::width == 5);
```

Comment:
If this code is no longer in use, could it be removed?

Reply:
@jheikkil @HaarigerHarald just wanted to ping you on this. 30 second clean-up here and I have seen enough of this emulator, and trust our validation enough to sign.

Reply:
This code is a placeholder, for  the future implementation of GTT jets as they (to our knowledge) is not present in cmssw.
But outlines the format for the expected input

Reply:
Okay. Good enough for me then. Is this expected near time scale? Otherwise a note above it might not be out of place.